File: api/src/services/minio.ts
import { Client } from 'minio';
import { config } from '@shared/config';

export const minioClient = new Client({
  endPoint: config.minio.endpoint.split(':')[0],
  port: parseInt(config.minio.endpoint.split(':')[1] || '9000'),
  useSSL: config.minio.useSSL,
  accessKey: config.minio.accessKey,
  secretKey: config.minio.secretKey,
});

export async function ensureBucket() {
  const exists = await minioClient.bucketExists(config.minio.bucket);
  if (!exists) {
    await minioClient.makeBucket(config.minio.bucket, 'us-east-1');
  }
}

export async function uploadEmail(path: string, content: Buffer): Promise<string> {
  await ensureBucket();
  await minioClient.putObject(config.minio.bucket, path, content);
  return path;
}

export async function getEmail(path: string): Promise<Buffer> {
  const stream = await minioClient.getObject(config.minio.bucket, path);
  const chunks: Buffer[] = [];
  
  return new Promise((resolve, reject) => {
    stream.on('data', (chunk) => chunks.push(chunk));
    stream.on('end', () => resolve(Buffer.concat(chunks)));
    stream.on('error', reject);
  });
}

export async function deleteEmail(path: string): Promise<void> {
  await minioClient.removeObject(config.minio.bucket, path);
}

File: smtp/internal/storage/minio.go
package storage

import (
	"context"
	"io"
	"strings"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
	"github.com/mymail/smtp/internal/config"
)

type MinIO struct {
	client *minio.Client
	bucket string
}

func NewMinIO(cfg config.MinIOConfig) (*MinIO, error) {
	endpoint := cfg.Endpoint
	useSSL := cfg.UseSSL

	// Parse endpoint
	parts := strings.Split(endpoint, ":")
	host := parts[0]
	port := "9000"
	if len(parts) > 1 {
		port = parts[1]
	}

	client, err := minio.New(host+":"+port, &minio.Options{
		Creds:  credentials.NewStaticV4(cfg.AccessKey, cfg.SecretKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		return nil, err
	}

	ctx := context.Background()
	exists, err := client.BucketExists(ctx, cfg.Bucket)
	if err != nil {
		return nil, err
	}
	if !exists {
		err = client.MakeBucket(ctx, cfg.Bucket, minio.MakeBucketOptions{})
		if err != nil {
			return nil, err
		}
	}

	return &MinIO{
		client: client,
		bucket: cfg.Bucket,
	}, nil
}

func (m *MinIO) Upload(ctx context.Context, path string, reader io.Reader, size int64) error {
	_, err := m.client.PutObject(ctx, m.bucket, path, reader, size, minio.PutObjectOptions{
		ContentType: "message/rfc822",
	})
	return err
}

// UploadStream streams data directly to MinIO without requiring a known size upfront
// This is used for streaming email uploads where we don't know the total size
func (m *MinIO) UploadStream(ctx context.Context, path string, reader io.Reader) error {
	// Use -1 for size to indicate unknown size, MinIO will handle streaming
	_, err := m.client.PutObject(ctx, m.bucket, path, reader, -1, minio.PutObjectOptions{
		ContentType: "message/rfc822",
	})
	return err
}

func (m *MinIO) Get(ctx context.Context, path string) (io.Reader, error) {
	obj, err := m.client.GetObject(ctx, m.bucket, path, minio.GetObjectOptions{})
	if err != nil {
		return nil, err
	}
	return obj, nil
}

func (m *MinIO) Delete(ctx context.Context, path string) error {
	return m.client.RemoveObject(ctx, m.bucket, path, minio.RemoveObjectOptions{})
}
